File: 2024/day17/clojure/solution.clj
#!/usr/bin/env clojure -M

;; Day 17: Chronospatial Computer - 3-bit VM emulator

(ns solution
  (:require [clojure.string :as str]))

(defn parse-input [text]
  "Parse registers and program from input."
  (let [lines (str/split-lines (str/trim text))
        a (Long/parseLong (second (re-find #"Register A: (\d+)" (nth lines 0))))
        b (Long/parseLong (second (re-find #"Register B: (\d+)" (nth lines 1))))
        c (Long/parseLong (second (re-find #"Register C: (\d+)" (nth lines 2))))
        program-str (second (re-find #"Program: ([\d,]+)" (nth lines 4)))
        program (vec (map #(Long/parseLong %) (str/split program-str #",")))]
    {:a a :b b :c c :program program}))

(defn combo-operand [operand a b c]
  "Get combo operand value."
  (case operand
    0 0
    1 1
    2 2
    3 3
    4 a
    5 b
    6 c
    (throw (ex-info "Invalid combo operand" {:operand operand}))))

(defn run-program [a b c program]
  "Execute the 3-bit computer program and return output."
  (let [program-len (count program)]
    (loop [ip 0
           a a
           b b
           c c
           output []]
      (if (>= ip program-len)
        output
        (let [opcode (nth program ip)
              operand (nth program (inc ip))]
          (case opcode
            ;; adv - A = A >> combo
            0 (let [shift (combo-operand operand a b c)]
                (recur (+ ip 2) (bit-shift-right a shift) b c output))

            ;; bxl - B = B XOR literal
            1 (recur (+ ip 2) a (bit-xor b operand) c output)

            ;; bst - B = combo % 8
            2 (let [val (bit-and (combo-operand operand a b c) 7)]
                (recur (+ ip 2) a val c output))

            ;; jnz - jump if A != 0
            3 (if (not= a 0)
                (recur operand a b c output)
                (recur (+ ip 2) a b c output))

            ;; bxc - B = B XOR C (ignores operand)
            4 (recur (+ ip 2) a (bit-xor b c) c output)

            ;; out - output combo % 8
            5 (let [val (bit-and (combo-operand operand a b c) 7)]
                (recur (+ ip 2) a b c (conj output val)))

            ;; bdv - B = A >> combo
            6 (let [shift (combo-operand operand a b c)]
                (recur (+ ip 2) a (bit-shift-right a shift) c output))

            ;; cdv - C = A >> combo
            7 (let [shift (combo-operand operand a b c)]
                (recur (+ ip 2) a b (bit-shift-right a shift) output))))))))

(defn part1 [text]
  "Run the program and return comma-separated output."
  (let [{:keys [a b c program]} (parse-input text)
        output (run-program a b c program)]
    (str/join "," output)))

(defn part2 [text]
  "Find initial A value that makes program output itself."
  (let [{:keys [b c program]} (parse-input text)
        program-len (count program)]

    ;; Work backwards from the last digit - build A 3 bits at a time
    (letfn [(search [target-idx current-a]
              (if (< target-idx 0)
                current-a
                ;; Try all 8 possible 3-bit values for this position
                (loop [bits 0]
                  (if (>= bits 8)
                    nil
                    (let [candidate-a (bit-or (bit-shift-left current-a 3) bits)]
                      ;; A can't be 0 at start (would halt immediately without output)
                      (if (and (= candidate-a 0) (= target-idx (dec program-len)))
                        (recur (inc bits))
                        (let [output (run-program candidate-a b c program)
                              expected (subvec program target-idx)]
                          (if (= output expected)
                            (if-let [result (search (dec target-idx) candidate-a)]
                              result
                              (recur (inc bits)))
                            (recur (inc bits))))))))))]
      (search (dec program-len) 0))))

(defn -main []
  (let [input-path (str (System/getProperty "user.dir") "/../input.txt")
        text (slurp input-path)]
    (println "Part 1:" (part1 text))
    (println "Part 2:" (part2 text))))

(-main)

File: 2024/day17/php/solution.php
<?php
/**
 * Day 17: Chronospatial Computer - 3-bit VM emulator
 */

/**
 * Parse input to extract registers and program
 */
function parseInput(string $text): array {
    $lines = explode("\n", trim($text));
    preg_match('/Register A: (\d+)/', $lines[0], $matchA);
    preg_match('/Register B: (\d+)/', $lines[1], $matchB);
    preg_match('/Register C: (\d+)/', $lines[2], $matchC);
    preg_match('/Program: ([\d,]+)/', $lines[4], $matchProg);

    $a = (int)$matchA[1];
    $b = (int)$matchB[1];
    $c = (int)$matchC[1];
    $program = array_map('intval', explode(',', $matchProg[1]));

    return [$a, $b, $c, $program];
}

/**
 * Execute the 3-bit computer program and return output
 */
function runProgram(int $a, int $b, int $c, array $program): array {
    $ip = 0;
    $output = [];
    $len = count($program);

    // Combo operand helper
    $combo = function(int $operand) use (&$a, &$b, &$c): int {
        return match($operand) {
            0, 1, 2, 3 => $operand,
            4 => $a,
            5 => $b,
            6 => $c,
            default => throw new InvalidArgumentException("Invalid combo operand: $operand"),
        };
    };

    while ($ip < $len) {
        $opcode = $program[$ip];
        $operand = $program[$ip + 1];

        switch ($opcode) {
            case 0: // adv - A = A >> combo
                $a = $a >> $combo($operand);
                break;
            case 1: // bxl - B = B XOR literal
                $b = $b ^ $operand;
                break;
            case 2: // bst - B = combo % 8
                $b = $combo($operand) & 7;
                break;
            case 3: // jnz - jump if A != 0
                if ($a !== 0) {
                    $ip = $operand;
                    continue 2;
                }
                break;
            case 4: // bxc - B = B XOR C
                $b = $b ^ $c;
                break;
            case 5: // out - output combo % 8
                $output[] = $combo($operand) & 7;
                break;
            case 6: // bdv - B = A >> combo
                $b = $a >> $combo($operand);
                break;
            case 7: // cdv - C = A >> combo
                $c = $a >> $combo($operand);
                break;
        }

        $ip += 2;
    }

    return $output;
}

/**
 * Part 1: Run the program and return comma-separated output
 */
function part1(string $text): string {
    [$a, $b, $c, $program] = parseInput($text);
    $output = runProgram($a, $b, $c, $program);
    return implode(',', $output);
}

/**
 * Part 2: Find initial A value that makes program output itself
 * Work backwards from the last digit, building A 3 bits at a time.
 */
function part2(string $text): int {
    [, $b, $c, $program] = parseInput($text);
    $progLen = count($program);

    // Recursive search for A value
    $search = function(int $targetIdx, int $currentA) use (&$search, $b, $c, $program, $progLen): ?int {
        if ($targetIdx < 0) {
            return $currentA;
        }

        // Try all 8 possible 3-bit values for this position
        for ($bits = 0; $bits < 8; $bits++) {
            $candidateA = ($currentA << 3) | $bits;

            // A can't be 0 at start (would halt immediately without output)
            if ($candidateA === 0 && $targetIdx === $progLen - 1) {
                continue;
            }

            $output = runProgram($candidateA, $b, $c, $program);

            // Check if output matches the suffix of the program
            $expected = array_slice($program, $targetIdx);
            if ($output === $expected) {
                $result = $search($targetIdx - 1, $candidateA);
                if ($result !== null) {
                    return $result;
                }
            }
        }

        return null;
    };

    return $search($progLen - 1, 0);
}

// Main
$inputPath = dirname(__DIR__) . '/input.txt';
$text = file_get_contents($inputPath);

echo "Part 1: " . part1($text) . "\n";
echo "Part 2: " . part2($text) . "\n";
